src/lib/isElementTypeOfComponent.ts

```diff
@@ -0,0 +1,28 @@
+import { Props } from '../../types/utils'
+
+const getComponentName = (component: React.ReactType) => {
+  switch (typeof component) {
+    case 'function':
+      return component.displayName || component.name || null // unknown
+    case 'string':
+      return component
+    default:
+      return null // unknown
+  }
+}
+
+const isElementTypeOfComponent = (
+  element: React.ReactElement<Props>,
+  component: React.ReactType,
+): boolean => {
+  const componentNameOfElement = getComponentName(element.type)
+  const componentName = getComponentName(component)
+
+  return (
+    componentNameOfElement !== null &&
+    componentName !== null &&
+    componentNameOfElement === componentName
+  )
+}
```

Comment:
Can you clarify why we need this function at all?

```jsx
const SFC = (props) => React.createElement('div', props)
class Classy extends React.Component {
  render() {
    return null
  }
}

// 

const sfc = React.createElement(SFC)
const classy = React.createElement(Classy)

//

classy.type === Classy // true
sfc.type === SFC // true
```

The main issue that this will not work for HOCs, however your solution will fail, too :)

Reply:
@layershifter 
Obviously I tried function comparison first. I have a sandbox (https://codesandbox.io/s/oqwj254k16) where both solution work.
Unfortunately function comparison doesn't work in our codebase. It was surprising for me too :( that's why I needed this method..
![screenshot 2018-11-22 at 18 29 39](https://user-images.githubusercontent.com/5442794/48916964-c4398d80-ee84-11e8-8d36-741c7e15a52e.png)

As of HOC argument, we don't really care about that, it's not a relevant use case; if users pass HOCs to the factory it's their problem.

In case I'm missing something, please provided valid user scenarios to counter this.

Reply:
I didn't try, but I think that issue is caused by `react-hot-loader`, BTW they have own util:
https://github.com/gaearon/react-hot-loader/blob/master/README.md#arecomponentsequalcomponent1-component2

Can we check it? I'm not sure that the comparison by names is the best idea.

Reply:
I'll take a look for sure if we keep current implementation idea